### PR TITLE
AP-143 Autoupdate system (client, server, autoupdate package)

### PR DESCRIPTION
**Pull request by innwkd**

### CODE REVIEW

File: packages/autoupdate/autoupdate.go [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+				params.BinaryParams{Version: v.version},
+			)
+			if err != nil {
+				log.WithFields(log.Fields{"type": consts.AutoupdateError, "err": err}).Error("get binary data")
+				return err
+			}
+
+			err = ioutil.WriteFile(executablePath, build.Body, 0755)
+			if err != nil {
+				log.WithFields(log.Fields{"type": consts.AutoupdateError, "err": err}).Error("update executable file")
```
Comment (potehinre) [inline packages/autoupdate/autoupdate.go]: IOError

File: packages/autoupdate/autoupdate.go [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+}
+
+func (u *Updater) restart(executablePath string) error {
+	cmd := exec.Command(executablePath)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	err := cmd.Start()
+	if err != nil {
+		return err
```
Comment (potehinre) [inline packages/autoupdate/autoupdate.go:92]: Logging cmdError?

File: packages/autoupdate/autoupdate.go [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+
+	u.versions = versions
+	return nil
+}
+
+// migrate executes database migrations
+func (u *Updater) migrate() error {
+	err := model.ExecSchema()
+	if err != nil {
+		log.WithFields(log.Fields{"type": consts.MigrationError, "err": err}).Errorf("apply migrations")
```
Comment (potehinre) [inline packages/autoupdate/autoupdate.go]: DBError

File: packages/autoupdate/autoupdate.go [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+		if err != nil {
+			log.WithFields(log.Fields{"type": consts.AutoupdateError, "err": err}).Error("parse version")
+			return nil, err
+		}
+		if ver.GreaterThan(appVer) {
+			neededVersions = append(neededVersions, &updateVersion{
+				version:     v.String(),
+				blockNumber: v.StartBlock,
+			})
+			log.Infof("Update to version %s is available", v.Number)
```
Comment (potehinre) [inline packages/autoupdate/autoupdate.go]: Use WithFields{"version_number": v.Number}

File: packages/daylight/system/pid.go [hunk 5]
```diff
@@ -0,0 +1,50 @@
+package system
```
Comment (potehinre) [inline packages/daylight/system/pid.go:1]: I personally hate this stuff, thats not unix-way, pid file should contain only pid)

File: packages/model/migration_history.go [hunk 6]
```diff
@@ -1,23 +1,44 @@
...
 }
 
-// Create is creating record of model
-func (mh *MigrationHistory) Create() error {
-	return DBConn.Create(mh).Error
+// ApplyMigration executes database schema and writes migration history
+func (mh *MigrationHistory) ApplyMigration(version, query string) error {
+	err := DBConn.Exec(query).Error
+	if err != nil {
+		return err
```
Comment (potehinre) [inline packages/model/migration_history.go:40]: Logging?

File: tools/update_server/storage/bolt_storage_test.go [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+
+	rw, err := bs.GetVersionsList()
+	require.NoError(t, err)
+	assert.Equal(t, wr, rw)
+
+	cleanUpFile(t, spath)
+
+}
+
+//func TestBoltStorage_GetVersionsList(t *testing.T) {
```
Comment (potehinre) [inline tools/update_server/storage/bolt_storage_test.go]: Commented code are not allowed

**Review by potehinre — APPROVED** [review]

File: packages/autoupdate/autoupdate.go (hunk 2 above)
Comment (dchertkov) [reply packages/autoupdate/autoupdate.go:92]: This is a local method that is logged in line 74

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**